In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
dataset = load_boston()
print(dataset.feature_names)
print(dataset.data)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']
[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (

In [2]:
df = pd.DataFrame(dataset.data, columns=dataset.feature_names)
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [3]:
df['MEDV'] = dataset.target
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [5]:
print(df.isnull().sum())

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64


## *There are 13 features in the dataset, we do not want to use all of these features for training our model, because not all of them are relevant.  Instead, we want to choose those features that directly influence the result (that is, prices of houses) to train the model. For this, we can use the corr() function.* 
## *The corr() function computes the pairwise correlation of columns:* 

In [16]:
corr = df.corr()
print(corr)

             CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247  0.352734   
ZN      -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
B       -0.385064  0.175520 

## *A perfect negative correlation is represented by the value –1.00: a 0.00 indicates no correlation and a +1.00 indicates a perfect positive correlation.*

## *From the MEDV column above in the output, you can see that the RM and LSTAT features have high correlation factors (positive and negative correlations) with the MEDV, also PTRATIO has the 3rd highest correlation.* 

## *Below, instead of visually finding the high correlation variables we'll find the top 3 variables (MEDV is the 4th but not counted) programmatically:*

In [17]:
#---get the top 4 features that has the highest correlation---
print(df.corr().abs().nlargest(4, 'MEDV').index)
#---print the top 4 correlation values---
print(df.corr().abs().nlargest(4, 'MEDV').values[:,13])

Index(['MEDV', 'LSTAT', 'RM', 'PTRATIO'], dtype='object')
[1.         0.73766273 0.69535995 0.50778669]


## *Above you can see the 3 high correlation variables are LSTAT, RM, and PTRATIO.  MEDV also appears but can't be counted*

## *Below we'll add the 3 variables LSTAT, RM and PTRATIO to 'x'*

In [20]:
x = pd.DataFrame(np.c_[df['LSTAT'], df['RM'], df['PTRATIO']], columns = ['LSTAT','RM','PTRATIO'])
Y = df['MEDV']

In [21]:
print(x)

     LSTAT     RM  PTRATIO
0     4.98  6.575     15.3
1     9.14  6.421     17.8
2     4.03  7.185     17.8
3     2.94  6.998     18.7
4     5.33  7.147     18.7
..     ...    ...      ...
501   9.67  6.593     21.0
502   9.08  6.120     21.0
503   5.64  6.976     21.0
504   6.48  6.794     21.0
505   7.88  6.030     21.0

[506 rows x 3 columns]


In [22]:
print(Y)

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: MEDV, Length: 506, dtype: float64


## *Below we'll split the data into 70% for training and 30% for testing*

In [23]:
from sklearn.model_selection import train_test_split
x_train, x_test, Y_train, Y_test = train_test_split(x, Y, test_size = 0.3, random_state=5)

## *After the split, let’s print out the shape of the training sets below*

In [24]:
print(x_train.shape)
print(Y_train.shape)

(354, 3)
(354,)


## *This means that the x training set now has 354 rows and 3 columns, while the Y training set (which contains the label) has 354 rows and 1 column.*

## *Let’s also print out the testing set:*

In [25]:
print(x_test.shape)
print(Y_test.shape)

(152, 3)
(152,)


## *This time, the testing set has 152 rows shown above*

## *We are now ready to begin the training. As you learned from the previous chapter, you can use the LinearRegression class to perform linear regression. In this case, we will use it to train our model:*

In [26]:
from sklearn.linear_model import LinearRegression
ml = LinearRegression()
ml.fit(x_train, Y_train)

LinearRegression()

In [27]:
y_pred=ml.predict(x_test)
print(y_pred)

[36.39913166 27.64533804 25.9228963   2.03218835 31.6409958   8.37701905
 26.57287073 28.08708988 26.38667873 20.17705848 31.3772592  20.95631138
 22.7421773  29.76220579 26.1729643  20.43913398  1.10417436 22.52194772
 13.00153832 19.79284434  5.85499553 20.6778144  36.75882832 24.22607003
 32.01240414 12.31191055 24.89763155 23.05674045 24.67595944 27.96374404
 12.59196156 10.48403218 16.73787305 22.78938498 28.19168706 19.82900903
 26.83233442 12.14110561 41.11066208 31.84146986 18.69329382  1.61472673
 27.8963771  15.57555336 29.26411886 32.68199415 -0.85449729 17.34104273
 25.04158844 12.33635509 18.27362243 20.26424139 24.31570859 14.43232405
 22.34627612 28.08141238 34.57111793 18.3707646  30.44730165 22.45222597
 19.58155891 24.83413911 14.34929508 33.94472208 20.54516129 12.52444261
 20.44124488 25.74446541 21.87149221 19.63726043 17.88287432 25.66715322
 18.15715115 18.78248485 18.00754868 30.58001266 23.82199431 18.61707456
 33.56924249 17.60847136 20.16905061 40.64466433 18

## *Values for MEDV prediction below are LSTAT, RM and PTRATIO entered in that order.  Values are taken from row 1 of data*

In [28]:
ml.predict([[4.98,6.575,15.3]])

/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([31.52873317])

## *Above you can see the prediction(31.5287) for MEDV based on the 3 input variables*

## *For comparison to actual MEDV (row 1) below (24.0)*

In [29]:
df['MEDV'] = dataset.target
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2